### PR TITLE
Update breaking_bad.py

### PR DESCRIPTION
**Pull request by dong-jy**

### CODE REVIEW

File: string/breaking_bad.py [hunk 1]
```diff
@@ -35,8 +49,27 @@ def match_symbol(chemicals, symbols):
 
     return combined
 
+print(match_symbol(chemicals, symbols))
+
+
+# An improved version of above, get right output.
```
Comment (ankit167) [inline string/breaking_bad.py]: Good to mention the time complexity as well.

**Comment by dong-jy** [comment]
Not sure about the O(mn) time complexity. Please see comment in file.

File: string/breaking_bad.py [hunk 2]
```diff
@@ -16,8 +16,22 @@
...
+input:
+chemicals = ['Amazon', 'Microsoft', 'Google']
+symbols = ['I', 'Am', 'cro', 'Na', 'le', 'abc', 'o']  # add 'o' here
+
+expected:
+['[am]azon', 'mi[cro]soft', 'goog[le]']
+
+exact output:
+['[Am]azon', 'Mi[cro]soft', 'Goog[le]', 'Amaz[o]n', 'Micr[o]s[o]ft', 'G[o][o]gle']
 """
```
Comment (ankit167) [inline string/breaking_bad.py]: What is the output that should be printed?
The 'expected' or the 'exact_output'? Any reason for mentioning both?

File: string/breaking_bad.py [hunk 3]
```diff
@@ -35,8 +49,36 @@ def match_symbol(chemicals, symbols):
...
+def match_symbol_1(symbols, words):
+    import re
+    bracketed_list = []
+    for w in words:
+        longest_match=''
+        for s in symbols:
+            matchs = re.findall(s, w)
+            for m in matchs:
+                longest_match = m if len(longest_match) < len(m) else longest_match
+        bracketed.append(re.sub(longest_match, '[{}]'.format(longest_match), w))
```
Comment (ankit167) [inline string/breaking_bad.py]: typo- should be 'bracketed_list'

File: string/breaking_bad.py [hunk 4]
```diff
@@ -35,8 +49,36 @@ def match_symbol(chemicals, symbols):
...
+O(mn) time
+m = len(symbols), n = len(words)
+Assuming re.findall(s, w) is O(1) time.
+"""
+
+def match_symbol_1(symbols, words):
+    import re
+    bracketed_list = []
+    for w in words:
+        longest_match=''
```
Comment (ankit167) [inline string/breaking_bad.py]: Pls add spaces around operator

File: string/breaking_bad.py [hunk 5]
```diff
@@ -35,8 +49,36 @@ def match_symbol(chemicals, symbols):
...
+        for s in symbols:
+            matchs = re.findall(s, w)
+            for m in matchs:
+                longest_match = m if len(longest_match) < len(m) else longest_match
```
Comment (ankit167) [inline string/breaking_bad.py]: line too long (83 > 79 characters)

File: string/breaking_bad.py [hunk 6]
```diff
@@ -35,8 +49,36 @@ def match_symbol(chemicals, symbols):
...
 
+print(match_symbol(chemicals, symbols))
+
+
+
+"""
+An improved version of above, get right output.
+
+symbols = ['I', 'Am', 'cro', 'Na', 'le', 'abc', 'o']
+words = ['Amazon', 'Microsoft', 'Google']
```
Comment (ankit167) [inline string/breaking_bad.py]: Pls maintain common list names throughout the code. I could see the 'chemicals' denoting the name of the list in match_symbols()

**Comment by ankit167** [comment]
@dong-jy want to understand, how this code is algorithmically better than match_symbols(). There are at least two nested loops being iterated in both the functions.

**Comment by dong-jy** [comment]
@ankit167 The new one **match_symbol_1()** is a naive approach that just output right right result.
The original **match_symbols()** function did wrong in inputs such as

 input:
 words = ['Amazon', 'Microsoft', 'Google']
 symbols = ['I', 'Am', 'cro', 'Na', 'le', 'abc', 'o'] 

 Exact putput:
 ['[Am]azon', 'Mi[cro]soft', 'Goog[le]', 'Amaz[o]n', 'Micr[o]s[o]ft', 'G[o][o]gle']

 Correct  answer should be:
 ['[Am]azon', 'Mi[cro]soft', 'Goog[le]']
 
 **match_symbols()** function's exact output  is wrong.
Because we should only substitute the longest matched sub-string in the word, for example choose word **'Google'** as word, then **'le'**, **'o'** both match 'Google', but we should only produce ['Goog[le]'], but not [ 'Goog[le]', 'G[o][o]gle']

File: string/breaking_bad.py (hunk 2 above)
Comment (dong-jy) [inline string/breaking_bad.py]: exact output:
['[Am]azon', 'Mi[cro]soft', 'Goog[le]', 'Amaz[o]n', 'Micr[o]s[o]ft', 'G[o][o]gle']
is printed by **match_symbols()**

File: string/breaking_bad.py (hunk 3 above)
Comment (dong-jy) [inline string/breaking_bad.py]: I'd revise it

File: string/breaking_bad.py (hunk 4 above)
Comment (dong-jy) [inline string/breaking_bad.py]: Yes

File: string/breaking_bad.py (hunk 6 above)
Comment (dong-jy) [inline string/breaking_bad.py]: Ok

**Comment by dong-jy** [comment]
Should I delete the previous match_symbols function?

**Comment by ankit167** [comment]
As you mentioned, only one version of the output looks correct, i.e. the longest matching substring. I'd suggest lets keep the new function- **match_symbol_1()** (rename it to match_symbol) and delete the original match_symbol() function. Pls edit the comments as well accordingly.

**Comment by dong-jy** [comment]
OK

**Comment by dong-jy** [comment]
OK

**Comment by outlawmose** [comment]
????

**Comment by outlawmose** [comment]
Approved

**Comment by dong-jy** [comment]
Revised

**Comment by keon** [comment]
@dong-jy thanks!

**Pull request merged**